_posts/2015-09-16-off-heap-memory.md

```diff
@@ -0,0 +1,875 @@
+
+
+## Understanding the JIT and tuning the implementation
+
+The `MemorySegment` was (before our change) a standalone class, it was *final* (had no subclasses). Via *Class Hierarchy Analysis (CHA)*, the JIT compiler was able to determine that all of the accessor method calls go to one specific implementation. That way, all method calls can be perfectly de-virtualized and inlined, which is essential to performance, and the basis for all further optimizations (like vectorization of the calling loop).
+
+With two different memory segments loaded at the same time, the JIT compiler cannot perform the same level of optimization any more, which results in a noticeable difference in performance: A slowdown of about 2.7 x in the following example:
+
+```
+Writing 100000 x 32768 bytes to 32768 bytes segment:
+
+HeapMemorySegment    (standalone) : 1,441 msecs
+OffHeapMemorySegment (standalone) : 1,628 msecs
+HeapMemorySegment    (subclass)   : 3,841 msecs
+OffHeapMemorySegment (subclass)   : 3,847 msecs
+```
+
+To get back to the original performance, we explored two approaches:
+
+### Approach 1: Make sure that only one memory segment implementation is ever loaded.
+
+We re-structured the code a bit to make sure that all places that produce long-lived and short-lived memory segments instantiate the same MemorySegment subclass (Heap- or Off-Heap segment). Using factories rather than directly instantiating the memory segment classes, this was straightforward.
+
+Experiments (see appendix) showed that the JIT compiler properly detects this (via hierarchy analysis) and that it can perform the same level of aggressive optimization as before, when there was only one `MemorySegment` class.
+
+
+### Approach 2: Write one segment that handles both heap and off-heap memory
+
+We created a class `HybridMemorySegment` which handles transparently both heap- and off-heap memory. It can be initialized either with a byte array (heap memory), or with a pointer to a memory region outside the heap (off-heap memory).
+
+Fortunately, there is a nice trick to do this without introducing code branches and specialized handling of the two different memory types. The trick is based on the way that the `sun.misc.Unsafe` methods interpret object references. To illustrate this, we take the method that gets a long integer from a memory position:
+
+```
+sun.misc.Unsafe.getLong(Object reference, long offset)
+```
+
+The method accepts an object reference, takes its memory address, and add the offset to obtain a pointer. It then fetches the eight bytes at the address pointed to and interprets them as a long integer. Since the method accepts *null* as the reference (and interprets it a *zero*) one can write a method that fetches a long integer seamlessly from heap and off-heap memory as follows:
```

Comment:
Let's hope it stays that way :) It's even documented:

http://www.docjar.com/docs/api/sun/misc/Unsafe.html#getInt%28Object,%20long%29
>Fetches a value from a given Java variable. More specifically, fetches a field or array element within the given object o at the given offset, or (if o is null) from the memory address whose numerical value is the given offset.